### ISSUE TITLE
[RFE] Add `DeviceSegmentedTopK`

### ISSUE THREAD

**Issue by elstehle** [NVIDIA/cccl#6391]
### Help Us Prioritize Top-K Algorithm Variants for `DeviceSegmentedTopK` (i.e., the segmented-problem, device-wide top-k algorithm)

We are planning the implementations for `DeviceSegmentedTopK`, and we’d like your input to help prioritize which variants we should focus on.

Please copy & paste the form below into a new comment and fill it out in order to describe **which top-k features or variants are most important to you**. Your responses will help us understand common use cases, preferred behaviors, and performance requirements, so we can prioritize the development of the most relevant versions.

<details>
<summary> 🗒️ Segmented Top-K Algorithm Specification Form </summary>

````markdown

# 🧮 Segmented Top-K Algorithm Specification Form

Please fill out the form below to describe which **Top-K algorithm parameters or variants** you need.  
Use the checkboxes (`[x]`) to select the features or behaviors you’d like supported (multiple selections are fine). You can also add clarifying notes inline.

**Use case:**  
```
If possible, briefly describe how or where you plan to use the top-K algorithm.
```

---



## ⚙️ Basic Algorithmic Properties

**Result ordering:**
- [ ] Ordered (top-k items are returned as a sorted sequence)
- [ ] Unordered (top-k items are returned in arbitrary order (i.e., a set of items))

**Stability:**
- [ ] Stable (retain the relative order from the input of selected items)
- [ ] Unstable (no order guarantee)

**Determinism:**
- [ ] Deterministic (bit-wise reproducible results)
- [ ] Non-deterministic (e.g., faster parallel variants)

---

## 🔢 Extended-K Behavior

**Extend `k` by all keys that compare equal?**
- [ ] Yes — include all elements with equal keys beyond `k`
- [ ] No — exactly `k` elements only

---

## 🧮 Algorithm Type

Choose the algorithmic approach:
- [ ] Radix-based
- [ ] Comparison-based

<details>
<summary>▶️ If Radix-based, please specify:</summary>

**Support for custom/arbitrary types?**
- [ ] Yes
- [ ] No

**Bit range to process:**  
**Would you like to be able to specify the bit range that top-k will be looking at?**
- [ ] Yes
- [ ] No
</details>

---

## 💾 Data Characteristics

**Common data types** (select all that apply):
- [ ] **Built-in / fundamental types**: standard integer and floating-point types (`uint8_t`, `int32_t`, `float`, `double`, etc.)
- [ ] **Extended floating-point types**: fp16, bfloat16, ...
- [ ] **Custom data types**
- [ ] **Specific subset only**: I primarily care about a few specific types: ☐ (please list them)


---

## 📏 Segment-specific parameters versus global parameters

Which of the following parameters vary across segments (e.g., for the first segment, I need the top-5 for the second segment I need the top-10 items) and which of these parameters are the same for each segment (e.g., "I want the top-5 items for each segment)

Please check the parameters that vary across different segments:
- [ ] Segment size
- [ ] `k`
- [ ] Result ordering
- [ ] Stability
- [ ] Determinism
- [ ] Radix-based versus comparison-bassed
- [ ] Bit range to process

## 📏 Typical Problem Sizes

**Common number of total input elements (i.e., the sum of segment sizes over all segments) (`N`):**
- [ ] 1 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [ ] Variable / depends on workload
- [ ] Custom: ☐ (please specify exact or range values)

**Common upper bound of segment size (i.e., the larges segment size across all segments):**
- [ ] 1 – 9
- [ ] 10 – 99
- [ ] 100 – 999
- [ ] 1 000 – 9 999
- [ ] 10 000 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [ ] Variable / depends on workload
- [ ] Custom: ☐ (please specify exact or range values)

**Common upper bound of the `k` values (i.e., the larges value of `k` across all segments)::**
- [ ] 1 – 9
- [ ] 10 – 99
- [ ] 100 – 999
- [ ] 1 000 – 9 999
- [ ] 10 000 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [ ] Fraction of `N` (e.g., `k = N / 10`)
- [ ] Variable / depends on workload
- [ ] Custom: ☐ (please specify)
````

</details>

_If you're interested in other Top-K algorithm variants such as `DeviceTopK`, `BlockTopK`, or `WarpTopK`, please refer to the corresponding issue below:_

- [📄 `DeviceTopK`: User requests & feature prioritization](https://github.com/NVIDIA/cccl/issues/6387)  
- [📄 `WarpTopK` / `BlockTopK`: User requests & feature prioritization](https://github.com/NVIDIA/cccl/issues/6389)

**Comment by davidwendt** [comment]
# 🧮 Segmented Top-K Algorithm Specification Form

**Use case:**  

The [RAPIDS cuDF](https://github.com/rapidsai/cudf) has [requirements](https://github.com/rapidsai/cudf/issues/16222) to support groupby top-k aggregation. This is essentially a process where group labels are identified (keys) and the corresponding values are aggregated (i.e. top-k per key set). The keys are normally determined using a sort or hash algorithm and the aggregation is generally some segmented operation on the values (usually CUB or custom kernels). So identifying keys is something handled by cuDF and the aggregated results are contiguous in segments.
---



## ⚙️ Basic Algorithmic Properties

**Result ordering:**
- [x] Ordered (top-k items are returned as a sorted sequence)
- [x] Unordered (top-k items are returned in arbitrary order (i.e., a set of items))
Either is fine though an ordered option would make it easier for test cases.

**Stability:**
- [x] Stable (retain the relative order from the input of selected items)
- [x] Unstable (no order guarantee)
Some upstream users of cuDF may require stable results.

**Determinism:**
- [ ] Deterministic (bit-wise reproducible results)
- [x] Non-deterministic (e.g., faster parallel variants)
cuDF currently has not had many requirements for deterministic results thus far.

---

## 🔢 Extended-K Behavior

**Extend `k` by all keys that compare equal?**
- [ ] Yes — include all elements with equal keys beyond `k`
- [x] No — exactly `k` elements only
What happens if there are less the k elements for a key?

---

## 🧮 Algorithm Type

Choose the algorithmic approach:
- [x] Radix-based
- [x] Comparison-based
Both of these would be desirable with also allowing a custom-based comparator.

<details>
<summary>▶️ If Radix-based, please specify:</summary>

**Support for custom/arbitrary types?**
- [ ] Yes
- [x] No
The CUB radix sort supports a decomposer which allows us to customize how NaN is handled for floats.
Does this include int128 and uint128?

**Bit range to process:**  
**Would you like to be able to specify the bit range that top-k will be looking at?**
- [ ] Yes
- [x] No
The bit-range is not used by cuDF
</details>

---

## 💾 Data Characteristics

**Common data types** (select all that apply):
- [x] **Built-in / fundamental types**: standard integer and floating-point types (`uint8_t`, `int32_t`, `float`, `double`, etc.)
- [ ] **Extended floating-point types**: fp16, bfloat16, ...
- [x] **Custom data types**
- [ ] **Specific subset only**: I primarily care about a few specific types: ☐ (please list them)


---

## 📏 Segment-specific parameters versus global parameters

Which of the following parameters vary across segments (e.g., for the first segment, I need the top-5 for the second segment I need the top-10 items) and which of these parameters are the same for each segment (e.g., "I want the top-5 items for each segment)

Please check the parameters that vary across different segments:
- [x] Segment size
- [ ] `k`
- [x] Result ordering
- [x] Stability
- [ ] Determinism
- [x] Radix-based versus comparison-based
- [ ] Bit range to process

## 📏 Typical Problem Sizes

**Common number of total input elements (i.e., the sum of segment sizes over all segments) (`N`):**
- [ ] 1 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [x] Variable / depends on workload
- [ ] Custom: ☐ (please specify exact or range values)

**Common upper bound of segment size (i.e., the larges segment size across all segments):**
- [ ] 1 – 9
- [ ] 10 – 99
- [ ] 100 – 999
- [ ] 1 000 – 9 999
- [ ] 10 000 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [x] Variable / depends on workload
- [ ] Custom: ☐ (please specify exact or range values)

**Common upper bound of the `k` values (i.e., the larges value of `k` across all segments)::**
- [ ] 1 – 9
- [ ] 10 – 99
- [ ] 100 – 999
- [ ] 1 000 – 9 999
- [ ] 10 000 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [x] Fraction of `N` (e.g., `k = N / 10`)
- [x] Variable / depends on workload
- [ ] Custom: ☐ (please specify)

Along with custom comparators and supporting custom types it would be helpful to be able to supply iterators for input parameters to minimize creating temporary buffers.

**Comment by ORippler** [comment]
# 🧮 Segmented Top-K Algorithm Specification Form

Please fill out the form below to describe which **Top-K algorithm parameters or variants** you need.  
Use the checkboxes (`[x]`) to select the features or behaviors you’d like supported (multiple selections are fine). You can also add clarifying notes inline.

**Use case:**  
### Pre-filtering of logits as part of token-sampling in ggml/llama.cpp (LLM inference). As soon as one batches multiple sequences, one has to go from Top-K to Segmented Top-K to yield said functionality.

---



## ⚙️ Basic Algorithmic Properties

**Result ordering:**
- [x] Ordered (top-k items are returned as a sorted sequence) <- not strictly necessary, as one can just sort the values afterwards if needed
- [x] Unordered (top-k items are returned in arbitrary order (i.e., a set of items))

**Stability:**
- [ ] Stable (retain the relative order from the input of selected items)
- [x] Unstable (no order guarantee)

**Determinism:**
- [ ] Deterministic (bit-wise reproducible results)
- [x] Non-deterministic (e.g., faster parallel variants)

---

## 🔢 Extended-K Behavior

**Extend `k` by all keys that compare equal?**
- [ ] Yes — include all elements with equal keys beyond `k`
- [x] No — exactly `k` elements only

---

## 🧮 Algorithm Type

Choose the algorithmic approach:
- [x] Radix-based
- [ ] Comparison-based

<details>
<summary>▶️ If Radix-based, please specify:</summary>

**Support for custom/arbitrary types?**
- [ ] Yes
- [x] No

**Bit range to process:**  
**Would you like to be able to specify the bit range that top-k will be looking at?**
- [ ] Yes
- [x] No <-- typically looking at the full data range
</details>

---

## 💾 Data Characteristics

**Common data types** (select all that apply):
- [x] **Built-in / fundamental types**: standard integer and floating-point types (`uint8_t`, `int32_t`, `float`, `double`, etc.)
- [x **Extended floating-point types**: fp16, bfloat16, ...
- [ ] **Custom data types**
- [ ] **Specific subset only**: I primarily care about a few specific types: ☐ (please list them)


---

## 📏 Segment-specific parameters versus global parameters

Which of the following parameters vary across segments (e.g., for the first segment, I need the top-5 for the second segment I need the top-10 items) and which of these parameters are the same for each segment (e.g., "I want the top-5 items for each segment)

Please check the parameters that vary across different segments:
- [ ] Segment size
- [x] `k`
- [ ] Result ordering
- [ ] Stability
- [ ] Determinism
- [ ] Radix-based versus comparison-bassed
- [ ] Bit range to process

## 📏 Typical Problem Sizes

**Common number of total input elements (i.e., the sum of segment sizes over all segments) (`N`):**
- [x] 1 – 99 999
- [x] 100 000 – 999 999
- [x] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [x] Variable / depends on workload <- the above ticks should roughly specify the ranges for now (may change based on the vocab size LLMs employ in future)
- [ ] Custom: ☐ (please specify exact or range values)

**Common upper bound of segment size (i.e., the larges segment size across all segments):**
- [ ] 1 – 9
- [ ] 10 – 99
- [ ] 100 – 999
- [ ] 1 000 – 9 999
- [ ] 10 000 – 99 999
- [x] 100 000 – 999 999
- [x] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [x] Variable / depends on workload <- the above ticks should roughly specify the ranges for now (may change based on the vocab size LLMs employ in future)
- [ ] Custom: ☐ (please specify exact or range values)

**Common upper bound of the `k` values (i.e., the larges value of `k` across all segments)::**
- [x] 1 – 9
- [x] 10 – 99
- [x] 100 – 999
- [x] 1 000 – 9 999
- [ ] 10 000 – 99 999
- [ ] 100 000 – 999 999
- [ ] 1 000 000 – 9 999 999
- [ ] 10 000 000 – 99 999 999
- [ ] 100 000 000 – 999 999 999
- [ ] ≥ 1 000 000 000
- [ ] Fraction of `N` (e.g., `k = N / 10`)
- [x] Variable / depends on workload <- bigger than 1024 unlikely
- [ ] Custom: ☐ (please specify)

**Comment by pifon2a** [comment]
@elstehle @danielfrg we are interested in migrating to CUB away from RAFT in XLA:GPU. Here is the list of kernels that we use in RAFT. https://github.com/openxla/xla/blob/2d62d5eb7e400bf871c5bbf9ba48ae9afbe3cadd/xla/backends/gpu/runtime/select_k_exec_raft.cc#L336

**Comment by jrhemstad** [comment]
Thanks @pifon2a we'd be happy to help!

Could you provide some info on the range of values that are important to your use cases? The ideal implementation depends heavily on the values that are important to your use case, so this information would help inform our prioritization and timeline. 

If you can, filling out the "Segmented Top-K Algorithm Specification Form" form at the [top of the issue](https://github.com/NVIDIA/cccl/issues/6391) would be ideal.

**Comment by pifon2a** [comment]
@jrhemstad Thank you! There is this [logic](https://github.com/openxla/xla/blob/e986182a6c1d07d7bf43631e34e87b6c852c3afd/xla/service/gpu/thunk_emitter.cc#L1195) to decide whether to use RAFT or not. It covers the current use case. It would be nice to migrate away from RAFT to CCCL first and then we can add more data types, sizes, etc.

**Comment by jrhemstad** [comment]
Thanks @pifon2a, this is very helpful.

It looks like you're still using a custom kernel for some parameter combinations as well. Would there be an interest in replacing that custom kernel with CUB's top-k as well assuming it meets or exceeds performance expectations?

Are there any benchmarks you can point us at to help us in evaluating if CUB's implementation meets your performance needs?

**Comment by pifon2a** [comment]
> Would there be an interest in replacing that custom kernel with CUB's top-k as well assuming it meets or exceeds performance expectations?

Yes! I am always happy to remove some code from XLA.

> Are there any benchmarks you can point us at to help us in evaluating if CUB's implementation meets your performance needs?

Not really. This RAFT integration was done, because one of the internal models needed it. Before RAFT we had the custom kernel for k < 16 for bfloat16 and float32. In all other cases, it was expanded to slice(sort), which definitely not the fastest way of doing topK. Later we introduced RAFT by comparing what we have (custom kernel + slice(sort) expansion) with what RAFT has. And that's how the [heuristic](https://github.com/openxla/xla/blob/2d62d5eb7e400bf871c5bbf9ba48ae9afbe3cadd/xla/backends/gpu/runtime/select_k_exec_raft.cc#L336) was made: by plotting the performance on h100 and b200 for different sizes and element types.

We are also interested in having float64 topk.

**Comment by apivovarov** [comment]
[RFC] The Stable TopK Adapter: Integrating Unstable TopK Kernels into XLA - https://github.com/openxla/xla/issues/43913

**Comment by apivovarov** [comment]
Decision Points for Using RAFT TopK in the XLA Codebase

1. topk_specializer SmallBufferOptimization - https://github.com/openxla/xla/blob/main/xla/backends/gpu/transforms/topk_specializer.cc#L80

2. ThunkEmitter::EmitTopKCustomCall - https://github.com/openxla/xla/blob/main/xla/service/gpu/thunk_emitter.cc#L1177

3. Once we decide to use raft we do another decision to select particular algo from raft gpu/runtime/select_k_exec_raft.cc
- f32 choose_select_k_algorithm<float> https://github.com/openxla/xla/blob/main/xla/backends/gpu/runtime/select_k_exec_raft.cc#L213
- bf16 choose_select_k_algorithm<nv_bfloat16> https://github.com/openxla/xla/blob/main/xla/backends/gpu/runtime/select_k_exec_raft.cc#L237
